src/Fixer/Basic/NumericLiteralSeparatorFixer.php

```diff
@@ -0,0 +1,182 @@
+use PhpCsFixer\FixerConfiguration\FixerOptionBuilder;
+use PhpCsFixer\FixerDefinition\CodeSample;
+use PhpCsFixer\FixerDefinition\FixerDefinition;
+use PhpCsFixer\FixerDefinition\FixerDefinitionInterface;
+use PhpCsFixer\Preg;
+use PhpCsFixer\Tokenizer\Token;
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * Let's you add underscores to numeric literals.
+ *
+ * Inspired by:
+ * - {@link https://github.com/kubawerlos/php-cs-fixer-custom-fixers/blob/main/src/Fixer/NumericLiteralSeparatorFixer.php}
+ * - {@link https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/rules/numeric-separators-style.js}
+ *
+ * @author Marvin Heilemann <marvin.heilemann+github@googlemail.com>
+ */
+final class NumericLiteralSeparatorFixer extends AbstractFixer implements ConfigurableFixerInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getDefinition(): FixerDefinitionInterface
+    {
+        return new FixerDefinition(
+            'Adds separators to numeric literals of any kind.',
+            [
+                new CodeSample("<?php\n\$integer = 12345678;\n\$octal = 0123456;\n\$binary = 0b00100100;\n\$hexadecimal = 0x3D458F4F;\n"),
+                new CodeSample(
+                    "<?php \$var = 24_40_21;\n",
+                    ['override_existing' => true]
+                ),
+            ],
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isCandidate(Tokens $tokens): bool
+    {
+        return $tokens->isAnyTokenKindsFound([T_DNUMBER, T_LNUMBER]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function createConfigurationDefinition(): FixerConfigurationResolverInterface
+    {
+        return new FixerConfigurationResolver([
+            (new FixerOptionBuilder('override_existing', 'Reformat literals already contain underscores.'))
+                ->setAllowedTypes(['bool'])
+                ->setDefault(false)
+                ->getOption(),
+        ]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
+    {
+        foreach ($tokens as $index => $token) {
+            if (!$token->isGivenKind([T_DNUMBER, T_LNUMBER])) {
+                continue;
+            }
+
+            $content = $token->getContent();
+
+            if ($this->configuration['override_existing']) {
+                $content = str_replace('_', '', $content);
+            } elseif (str_contains($content, '_')) {
+                // Keep already underscored literals untouched.
+                continue;
+            }
+
+            $newContent = $this->formatValue($content);
+
+            if ($content === $newContent) {
+                // Skip Token override if its the same content, like when it
+                // already got a valid literal separator structure.
+                continue;
+            }
+
+            $token = new Token([$token->getId(), $newContent]);
+
+            $tokens[$index] = $token;
```

Comment:
Unneeded assignment:

```suggestion
            $tokens[$index] = new Token([$token->getId(), $newContent]);
```